cgen: -autofree generates invalid C code for `g(f()!)`, where `f` is `fn f() !string { return 'abc' }`

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.4.0 63867d4.24d3718
OS: linux, Ubuntu 20.04.6 LTS
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i3-3225 CPU @ 3.30GHz

getwd: /v/vnew
vexe: /v/vnew/v
vexe mtime: 2023-07-21 10:17:40

vroot: OK, value: /v/vnew
VMODULES: OK, value: /home/delian/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.41.0
Git vroot status: weekly.2023.28-59-g96ff3ce8
.git/config present: true

CC version: cc (Ubuntu 10.3.0-1ubuntu1~20.04) 10.3.0
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3

```

**What did you do?**
`v -autofree x.v`
```v
fn f() !string {
	return 'abc'
}

fn g(s string) {
	println(s)
}

fn main() {
	g(f()!)
}

```

**What did you expect to see?**

A compiled program, even with -autofree


**What did you see instead?**
```
v -autofree x.v 
==================
/tmp/v_1000/x.1818340312026700634.tmp.c:12364: error: 'ror' undeclared
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
```

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -507,8 +507,11 @@ fn (mut c Checker) call_expr(mut node ast.CallExpr) ast.Type {
 	c.inside_fn_arg = old_inside_fn_arg
 	// autofree: mark args that have to be freed (after saving them in tmp exprs)
 	free_tmp_arg_vars := c.pref.autofree && !c.is_builtin_mod && node.args.len > 0
-		&& !node.args[0].typ.has_flag(.option) && !node.args[0].typ.has_flag(.result)
-	if free_tmp_arg_vars && !c.inside_const {
+		&& !c.inside_const && !node.args[0].typ.has_flag(.option)
+		&& !node.args[0].typ.has_flag(.result) && !(node.args[0].expr is ast.CallExpr
+		&& (node.args[0].expr.return_type.has_flag(.option)
+		|| node.args[0].expr.return_type.has_flag(.result)))
+	if free_tmp_arg_vars {
 		for i, arg in node.args {
 			if arg.typ != ast.string_type {
 				continue

--- a/vlib/v/slow_tests/valgrind/fn_call_result_arg.v
+++ b/vlib/v/slow_tests/valgrind/fn_call_result_arg.v
@@ -1,0 +1,11 @@
+fn f() !string {
+	return 'abc'
+}
+
+fn g(s string) {
+	println(s)
+}
+
+fn main() {
+	g(f()!)
+}

